test/forms/sa/SAUTRFormProviderSpec.scala

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2018 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package forms.sa
+
+import forms.behaviours.FormBehaviours
+import models.sa.SAUTR
+import models.{Field, Required}
+import org.scalacheck.Gen
+import org.scalacheck.Gen._
+import org.scalatest.prop.GeneratorDrivenPropertyChecks
+import play.api.data.Form
+import scala.util.Random.shuffle
+
+class SAUTRFormProviderSpec extends FormBehaviours with GeneratorDrivenPropertyChecks {
+
+  override val validData: Map[String, String] = Map("value" -> "0987654321")
+  override val form: Form[_] = new SAUTRFormProvider()()
+
+  private val utrLength = 10
+  private val min = 0
+  private val max = 9
+
+  val validGen: Gen[SAUTR] =
```

Comment:
Could a general Gen be created that is used to implement these 3 generators?
Something like
```
def saUtrGen[A](listGen: Gen[List[A]], pred: List[A] => Boolean = _ => true) = {
  for {
    list         <- listGen
    if pred(list)
    spaces <- listOf(const(' '))
    random = shuffle(list ++ spaces)
  } yield {
    SAUTR(random.mkString)
  }
}
```

Then the other gens would look like
```
val validGen = saUtrGen(listOfN(utrLength, choose(min, max)))

val invalidLengthGen = saUtrGen(listOf(choose(min, max)), _.length != utrLength && _.nonEmpty)

val invalidCharGen = saUtrGen(listOfN(utrLength, asciiChar), !_.contains(' '))
```